dg.go

```diff
@@ -1,315 +1,464 @@
+	}
+	n.status = status
+	if status == Waiting {
+		return nil // Don't propagate a waiting status.
+	}
+	// Propagate to outbound connections.
+	outboundConnections := n.dg.connectionsFromNode[n.ID()]
+	for outboundConnectionID := range outboundConnections {
+		err := n.dg.nodes[outboundConnectionID].dependencyResolved(n.ID(), status)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (n *node[NodeType]) Connect(nodeID string) error {
+	return n.dg.connectNodes(n.id, nodeID, AndDependency)
+}
+
+func (n *node[NodeType]) ConnectDependency(fromNodeID string, dependencyType DependencyType) error {
+	return n.dg.connectNodes(fromNodeID, n.id, dependencyType)
 }
 
-func (d *node[NodeType]) ID() string {
-    return d.id
+func (n *node[NodeType]) DisconnectInbound(fromNodeID string) error {
+	n.dg.lock.Lock()
+	defer n.dg.lock.Unlock()
+	if n.deleted {
+		return &ErrNodeDeleted{n.id}
+	}
+	if _, ok := n.dg.nodes[fromNodeID]; !ok {
+		return &ErrNodeNotFound{fromNodeID}
+	}
+	if _, ok := n.dg.connectionsToNode[n.id][fromNodeID]; !ok {
+		return &ErrConnectionDoesNotExist{n.id, fromNodeID}
+	}
+	delete(n.dg.connectionsToNode[n.id], fromNodeID)
+	delete(n.dg.connectionsFromNode[fromNodeID], n.id)
+	return nil
 }
 
-func (d *node[NodeType]) Item() NodeType {
-    return d.item
+func (n *node[NodeType]) DisconnectOutbound(toNodeID string) error {
+	n.dg.lock.Lock()
+	defer n.dg.lock.Unlock()
+	if n.deleted {
+		return &ErrNodeDeleted{n.id}
+	}
+	if _, ok := n.dg.nodes[toNodeID]; !ok {
+		return &ErrNodeNotFound{toNodeID}
+	}
+	if _, ok := n.dg.connectionsFromNode[n.id][toNodeID]; !ok {
+		return &ErrConnectionDoesNotExist{n.id, toNodeID}
+	}
+	delete(n.dg.connectionsFromNode[n.id], toNodeID)
+	delete(n.dg.connectionsToNode[toNodeID], n.id)
+	return nil
 }
 
-func (d *node[NodeType]) Connect(nodeID string) error {
-    d.dg.lock.Lock()
-    defer d.dg.lock.Unlock()
-    if d.deleted {
-        return &ErrNodeDeleted{
-            d.id,
-        }
-    }
-    if nodeID == d.id {
-        return &ErrCannotConnectToSelf{
-            nodeID,
-        }
-    }
-    if _, ok := d.dg.nodes[nodeID]; !ok {
-        return &ErrNodeNotFound{
-            nodeID,
-        }
-    }
-    if _, ok := d.dg.connectionsFromNode[d.id][nodeID]; ok {
-        return &ErrConnectionAlreadyExists{
-            d.id,
-            nodeID,
-        }
-    }
-    d.dg.connectionsFromNode[d.id][nodeID] = struct{}{}
-    d.dg.connectionsToNode[nodeID][d.id] = struct{}{}
-    return nil
+func (n *node[NodeType]) Remove() error {
+	n.dg.lock.Lock()
+	defer n.dg.lock.Unlock()
+	if n.deleted {
+		return &ErrNodeDeleted{n.id}
+	}
+	for toNodeID := range n.dg.connectionsFromNode[n.id] {
+		delete(n.dg.connectionsToNode[toNodeID], n.id)
+	}
+	delete(n.dg.connectionsFromNode, n.id)
+	for fromNodeID := range n.dg.connectionsToNode[n.id] {
+		delete(n.dg.connectionsFromNode[fromNodeID], n.id)
+	}
+	delete(n.dg.connectionsToNode, n.id)
+	delete(n.dg.nodes, n.id)
```

Comment:
It's above the loop rather than below, since the loop is a preface to deleting `connectionsToNode`. Yeah, it's possible that the flow would be a little easier to read if the three `delete` calls were together, possibly preceded by the two loops? But I don't think that's enough of an improvement to bother with. A *comment* about flushing references to this node from the two maps before deleting the maps might help, though.